qt5-packages.nix: consistently make all libraries with same qt5 version

FRidh opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Reduce the chance of mixing Qt versions by having Qt-based libraries in a separate package set.

Builds upon https://github.com/NixOS/nixpkgs/pull/101369.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).


## To do
- [ ] move `callPackage` into `qt-packages.nix`
- [ ] consider https://github.com/NixOS/nixpkgs/pull/105374 for cross-compilation. I do think all the `makeScope` usage for Qt and such itself is going to make this difficult
- [x] update docs
- [ ] rename `libsForQt5`

FRidh committed:
qt5-packages.nix: consistently make all libraries with same qt5 version

ttuegel commented on pkgs/top-level/qt-packages.nix:
```diff
@@ -0,0 +1,171 @@
+
+# Qt libraries set.
```
This says the file is for Qt libraries, but it includes all of `plasma5` and `kdeApplications`. Why isn't this all Qt packages? I think that would be a better solution to the version-inconsistency problem.

FRidh replied:
I am not sure whether we should use it only for libraries, or also applications. When I created this I aimed for libraries only.

The issue we have here seems similar to Python, however, somewhat simpler because we do not have to deal with various versions of libraries other than the versions of Qt. This is the reason Python applications are outside of `python-packages.nix`, so you can use different versions. But with Qt I *think* we do not need it, so then we can indeed put both applications and libraries in the same file.

doronbehar replied:
The distinction between an application and a library can't be guaranteed to be clear for ever. In #101369 I learned that some "applications" from `kdeApplications` were used as inputs to other. Hence I support the idea to put all qt applications there (docs should be updated as well).

ttuegel commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13318,15 +13318,7 @@ in
-      };
-    in
-      recurseIntoAttrs (makeOverridable mkFrameworks attrs);
+  inherit (libsForQt5) kdeFrameworks;
```
Exposing a default `kdeFrameworks` attribute at the top level is going to perpetuate the problem we already have with packages importing dependencies from the wrong set. All Qt packages that depend on KDE Frameworks must import those libraries with `libsForQt5*`, so why should this be here?

FRidh replied:
Right. I added this here for backwards-compatibility.

I think we should get rid of it, and link to each of the applications here, via `qt-packages.nix`. So, you would something like `inherit (libsForQt5) dolphin;`, which is like how it was before.

It means undoing use of `aliases.nix` for the applications, which I think in retrospect was an incorrect solution.

doronbehar replied:
I agree with @ttuegel on this one as well.

> It means undoing use of `aliases.nix` for the applications, which I think in retrospect was an incorrect solution.

I did it in order to avoid future usage of attributes to be used as inputs from the wider set, vs as inputs from the correct `libsForQt5*` set. It doesn't have to be that way if they'd all be part of qt5-packages.nix.

FRidh replied:
I have removed the top-level `kdeFrameworks`, `kdeApplications` and `plasma5`, however, I did add in return a `plasma5Packages` which maps to the version of `libsForQt5` that is used for `plasma5`, mainly for NixOS.

doronbehar commented:
Long term thinking: How about changing a bit the attributes layout to look like this (json5):

```json5
qt5 = {
  // Same as today
  qtbase = ..
  // Other qt modules
  libraries = {
    // All what was previously in libsForQt5
  },
  apps = {
    // the apps from qt5-applications.nix
  }
}
```

And so on, with `qt515` and `qt514` etc.

ttuegel commented:
> How about changing a bit the attributes layout to look like this (json5):
> 
> ```js
> qt5 = {
>   // Same as today
>   qtbase = ..
>   // Other qt modules
>   libraries = {
>     // All what was previously in libsForQt5
>   },
>   apps = {
>     // the apps from qt5-applications.nix
>   }
> }
> ```
> 
> And so on, with `qt515` and `qt514` etc.

I don't really like this beacuse, as you rightly pointed out, the distinction between "library" and "application" is flexible.

I suggest the following, which I think is the same as the situation for Python right now:

1. All packages that depend on Qt go into `qt-packages.nix`.
2. Packages in `qt-packages.nix` get their dependencies from that set _first_, so that they see a consistent set of dependencies.
3. In `all-packages.nix`, we have attributes for the entire `qt-packages.nix` set instantiated with different versions of Qt: `qtPackages5_14`, `qtPackages5_15`, etc. The default version will be `qtPackages5 = qtPackages5_yy`. Hydra will _not_ recurse into these sets.
4. For stand-alone packages (that is: applications) add an attribute to `all-packages.nix` like `konsole = qtPackages5.konsole`. If the package needs to be pinned to a certain version of Qt, the maintainer can do that, too.

I think this would fix the mixed-version problems we have. The one problem: I don't know how to make this work with PyQt packages; they would somehow need to live in two package sets at once. I guess we can leave PyQt in `pythonPackages` (built with all versions of Python) and pin it to a particular version of Qt because all Python packages that depend on Qt should get it through PyQt so there is no risk of mixing versions.

FRidh committed:
plasma5Packages: alias to the libsForQt5 used to build the plasma5 desktop

For in NixOS it is beneficial if both plasma5 and pam use the same Qt5
version. Because the plasma5 desktop may use a different version as the
default Qt5 version, we introduce plasma5Packages.

FRidh committed:
breeze-icons: take from plasma5Packages

FRidh commented:
For clarity I've kept `libsForQt5` for now. Once everything else is fine, we can do a simple replacement to the chosen name.

ttuegel commented on pkgs/applications/networking/browsers/firefox/wrapper.nix:
```diff
@@ -76,7 +76,7 @@ let
           ++ lib.optional (cfg.enableGnomeExtensions or false) chrome-gnome-shell
           ++ lib.optional (cfg.enableUgetIntegrator or false) uget-integrator
-          ++ lib.optional (cfg.enablePlasmaBrowserIntegration or false) plasma5.plasma-browser-integration
+          ++ lib.optional (cfg.enablePlasmaBrowserIntegration or false) plasma5Packages.plasma-browser-integration
```
This isn't directly related to this pull request, but would it be better to import `plasma-browser-integration` directly, and `inherit` it from `plasma5Packages` in `all-packages.nix`?

ttuegel commented:
@FRidh Looks good to me! :+1:

FRidh committed:
Qt docs: we now use qt5-packages.nix

FRidh commented:
> If what's here now is fine, I'll move over more callPackage lines into qt-packages.nix.

Did not do this eventually and this is now a follow-up task. Resolves the merge conflicts and merged this to avoid getting more conflicts.

SuperSandro2000 commented:
This PR broke eval on staging because fcitx-qt5 is no longer available. https://github.com/NixOS/nixpkgs/commit/2e9c639cb2a7ddf74956da2a945f8c9420f51297#diff-ab5748dc9567516fefba8344056b51ec1866adeace380f46e58a7af3d619ea22L15904

cole-h commented:
Rather, `fcitx5-qt`; https://github.com/NixOS/nixpkgs/pull/108950 and https://gist.github.com/8b2da452fad700c6e82220ae8039cca5.

FRidh commented:
Thanks. Fixed with 05ca995e31230a28306bf7d2f33ece4e538e7ca7.

anund commented:
Is there an additional change to turn `error: undefined variable 'kdeApplications' at` into an informative message? Looks like the equivalent for `kdeApplications.dolphin-plugins` is now `plasma5Packages.dolphin-plugins`.

anund commented:
or `plasma5Packages.kdeApplications.dolphin-plugins`.

jorsn commented:
> The one problem: I don't know how to make this work with PyQt packages; they would somehow need to live in two package sets at once. I guess we can leave PyQt in `pythonPackages` (built with all versions of Python) and pin it to a particular version of Qt because all Python packages that depend on Qt should get it through PyQt so there is no risk of mixing versions.
> -- @ttuegel (https://github.com/NixOS/nixpkgs/pull/102168#issuecomment-739120267)

If you have scope `A` with subscopes `B1`, `B2` and `C` and you want to change the version of `B1` to `B2` for use in `C`, you have to override `B1` in `A` i.e. set `A.C = (A.extend (self: super: { B1 = A.B2; })).C` to avoid inconsistent versions at some level. So PyQt5 needs to be pinned to the same qt5 that the whole python closure uses. If python used another qt version than the tree-wide default this would probably lead to the same practically unmanageable mess we had in September (see #99956).
